### ISSUE TITLE
Lensing task with the posterior distribution from the power spectrum analysis as proposal distribution

### ISSUE THREAD

**Issue by Justinezgh** [Justinezgh/SBI-Diff-Simulator#33]
**Goal**: Infer the posterior from log normal mass map (green contour)
![image](https://user-images.githubusercontent.com/72011736/210013987-1c2e9d50-2a8f-4130-bfa3-7f846232e8ce.png)
we would like to use the posterior of the power spectrum analysis as a proposal distribution to compress the mass map and then perform the inference. 

**Problem**: it's hard to compress the mass map into 2 summary statistics and get the green contour.

**Comment by Justinezgh** [comment]
Here I compress using 
- mse loss 
- proposal distribution: prior
- 100 000 mass map simulations 
- no data augmentation

This is the training and validation loss:
<img width="964" alt="image" src="https://user-images.githubusercontent.com/72011736/210014587-d9373ee4-26d3-45f3-bac8-02dc4ad47cd4.png">

Predict results on the data train:
![image](https://user-images.githubusercontent.com/72011736/210015150-948de2bd-ea04-47fa-ae72-c62c6c5d6dd3.png)

Predict results on the data test:
![image](https://user-images.githubusercontent.com/72011736/210016457-1a9cde9c-c599-4afe-9762-a16fc309e175.png)

The loss (nll) for the inference part.
<img width="964" alt="image" src="https://user-images.githubusercontent.com/72011736/210015229-ee4e66f5-01a2-4d7a-95ed-f4df6878f532.png">

![image](https://user-images.githubusercontent.com/72011736/210015273-6c461235-4e04-4a7f-aedc-4d1427c76597.png)

I trained the nf here on the same dataset used for compression.

**Comment by Justinezgh** [comment]
Here I compress using 
- mse loss 
- proposal distribution: prior
- 100 000 mass map simulations 
- **data augmentation**
```python 
def augmentation(example):
      x = example['simulation'] 
      x = tf.image.random_flip_left_right(x)
      x = tf.image.random_flip_up_down(x)

      return {'simulation': x, 'theta': example['theta'], 'score': example['score']}
```
In red, training and validation loss:
<img width="883" alt="image" src="https://user-images.githubusercontent.com/72011736/210135821-cc475cbd-0f1e-4da7-b6e3-933536358f69.png">

Predict results on the data train:
![image](https://user-images.githubusercontent.com/72011736/210136009-22aeb64e-75a7-4808-877d-336b1a112873.png)


Predict results on the data test:
![image](https://user-images.githubusercontent.com/72011736/210136043-2cb210da-5d39-4fd4-afc8-f85e9dff5064.png)


The loss (nll) for the inference part.
<img width="887" alt="image" src="https://user-images.githubusercontent.com/72011736/210153546-9b1c9242-1903-4d17-9a71-0fb7d5fbb438.png">

![image](https://user-images.githubusercontent.com/72011736/210153549-ac4f13fc-acd3-4602-a13e-49bb0db3ec95.png)


I trained the nf here on the same dataset used for compression but without augmentation.

**Comment by Justinezgh** [comment]
Here I compress using 
- mse loss 
- proposal distribution: prior
- 100 000 mass map simulations without noise
- **data augmentation**
```python 
def augmentation_with_noise(example):

      N=128
      map_size=5
      gal_per_arcmin2=30
      sigma_e=0.2
      pix_area = (map_size * 60 / N)**2  

      x = example['simulation'] + tf.random.normal(shape=(N, N), 
                                                  stddev=sigma_e/jnp.sqrt(gal_per_arcmin2 * pix_area))
      x = tf.image.random_flip_left_right(x)
      x = tf.image.random_flip_up_down(x)

      return {'simulation': x, 'theta': example['theta'], 'score': example['score']}
```
Training and validation loss:
<img width="965" alt="image" src="https://user-images.githubusercontent.com/72011736/210539111-aee2e0da-f960-4a9a-80fc-d8d2fdc02b76.png">

Predict results on the data test ( the 100 000 sim with noise added after):
![image](https://user-images.githubusercontent.com/72011736/210539995-504b7543-8553-4131-98ae-44c9121f8fa3.png)


Predict results on the data test ( the 100 000 sim with noise):
![image](https://user-images.githubusercontent.com/72011736/210539198-a7ffa119-4c9b-4634-96e6-ae399f518a9f.png)



The loss (nll) for the inference part.
<img width="705" alt="image" src="https://user-images.githubusercontent.com/72011736/210539639-958ef18e-a007-4308-a1da-ac3248d8b3bd.png">

![image](https://user-images.githubusercontent.com/72011736/210539691-e7443290-b63d-4951-ae7a-b158a3db14d0.png)


I trained the nf here on the 100 000 sim with noise without augmentation.

**Comment by Justinezgh** [comment]
averaged over 128 mass map
![image](https://user-images.githubusercontent.com/72011736/211620044-fa6cb6d4-5e63-48c1-9475-68c6d6af6aab.png)

averaged over 64 mass map 
![image](https://user-images.githubusercontent.com/72011736/211620150-271add15-9152-473f-811a-026669c25aa5.png)

purple correspond to simulations obtain directly from the simulator [as you can see here](https://github.com/Justinezgh/SBI-Diff-Simulator/blob/e44d891f8f564772c3239ab547de3957ef6f277e/sbids/tasks/lensinglognormal.py#L83) while red correspond to the field [here](https://github.com/Justinezgh/SBI-Diff-Simulator/blob/e44d891f8f564772c3239ab547de3957ef6f277e/sbids/tasks/lensinglognormal.py#L85) + noise added with data augmentation as follow: 

```python
def augmentation_with_noise(example):

      N=128
      map_size=5
      gal_per_arcmin2=30
      sigma_e=0.2
      pix_area = (map_size * 60 / N)**2  

      x = example['simulation'] + tf.random.normal(shape=(N, N), 
                                                  stddev=sigma_e/jnp.sqrt(gal_per_arcmin2 * pix_area))
      x = tf.image.random_flip_left_right(x)
      x = tf.image.random_flip_up_down(x)

      return {'simulation': x, 'theta': example['theta'], 'score': example['score']}
```

**Comment by Justinezgh** [comment]
Here I compress using 
- mse loss 
- proposal distribution: prior
- 100 000 mass map simulations without noise
- **data augmentation**
```python 
tf.random.set_seed(1)
                           
def augmentation_with_noise(example):

      N=128
      map_size=5
      gal_per_arcmin2=30
      sigma_e=0.2
      pix_area = (map_size * 60 / N)**2  

      x = tf.expand_dims(example['simulation'], axis = -1) + tf.random.normal(shape=(N, N, 1), 
                                             stddev=sigma_e/tf.math.sqrt(gal_per_arcmin2 * pix_area))

    
      x = tf.image.random_flip_left_right(x)
      x = tf.image.random_flip_up_down(x)

      return {'simulation': x, 'theta': example['theta'], 'score': example['score']}
```
and it is used as follow: 


``` python 
ds = tfds.load('LensingLogNormalDataset/toy_model_without_noise', 
               split='train', 
               data_dir = 'tensorflow_dataset')


ds = ds.repeat()
ds = ds.shuffle(1000)
ds = ds.map(augmentation_with_noise)
ds = ds.batch(128)
ds = ds.prefetch(tf.data.experimental.AUTOTUNE)
ds_train = iter(tfds.as_numpy(ds))
```
Training and validation loss:
![image](https://user-images.githubusercontent.com/72011736/213446333-daf9048e-1374-482e-8622-a51edc66b7d3.png)


Predict results on the data test ( the 100 000 sim with noise):
![image](https://user-images.githubusercontent.com/72011736/213466732-9fe74198-d5ed-4a6e-ae86-d805a78ee958.png)


Predict results on the data train ( the 100 000 sim with noise added after):
![image](https://user-images.githubusercontent.com/72011736/213466268-446566e2-9d35-4ccb-b256-050eeadf6425.png)


The loss (nll) for the sbi part.
![image](https://user-images.githubusercontent.com/72011736/213465927-405c59cb-b81f-4b70-aa65-abb4d6389332.png)


![image](https://user-images.githubusercontent.com/72011736/213465708-1e6c2c92-fb43-4073-ae49-c2bea129c4e7.png)


I trained the nf here on the 100 000 sim with noise without augmentation.

**Comment by EiffL** [comment]
hummm..... NN not doing better than power spectrum :-|

**Comment by Justinezgh** [comment]
Nop I'll try with mutual information now.

**Comment by Justinezgh** [comment]
With mutual information and the previous setup: 
![image](https://user-images.githubusercontent.com/72011736/214511725-2dbf4369-d423-4657-9107-451dc7960001.png)

So finally have the compressor. Next steps: perform SBI using the power spectrum contour as a proposal and show that using gradients we can reduce the number of simulations needed.

**Comment by Justinezgh** [comment]
Here the proposal is the power spectrum so the posterior is obtained by calculating nf_prediction * prior / proposal ( which is not so simple to compute --')

The contour (orange) I got with 100 000 simulations and no score (just to test) compare to the full field one (hist2d):
![image](https://user-images.githubusercontent.com/72011736/214512507-e3d01431-ca2a-4a12-aeda-99e892891943.png)
